packages/@aws-cdk/aws-glue/lib/job.ts

```diff
@@ -0,0 +1,872 @@
+   */
+  public onFailure(id?: string, options: events.OnEventOptions = {}): events.Rule {
+    return this.rule(JobState.FAILED, id, options);
+  }
+
+  /**
+   * Return a CloudWatch Event Rule matching TIMEOUT state.
+   *
+   * If no id or options are provided, the created rule is cached. Later no-args calls with retrieves from cache but ones with args.
+   * Later calls with args lead to the creation of a new Rule
+   *
+   * @param id optional construct id. default is TIMEOUTRule
+   * @param options optional event options. default is {}
+   */
+  public onTimeout(id?: string, options: events.OnEventOptions = {}): events.Rule {
+    return this.rule(JobState.TIMEOUT, id, options);
+  }
+
+  /**
+   * Create a CloudWatch metric.
+   *
+   * @param metricName name of the metric typically prefixed with `glue.driver.`, `glue.<executorId>.` or `glue.ALL.`.
+   * @param jobRunId a dimension that filters for metrics of a specific JobRun ID, or `ALL`.
+   * @param type the metric type.
+   * @param props metric options.
+   *
+   * @see https://docs.aws.amazon.com/glue/latest/dg/monitoring-awsglue-with-cloudwatch-metrics.html
+   */
+  public metric(metricName: string, jobRunId: string, type: MetricType, props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return new cloudwatch.Metric({
+      metricName,
+      namespace: 'Glue',
+      dimensions: {
+        JobName: this.jobName,
+        JobRunId: jobRunId,
+        Type: type,
+      },
+      ...props,
+    }).attachTo(this);
+  }
+
+  /**
+   * Return a CloudWatch Metric indicating job success.
+   *
+   * This metric is based on the Rule returned by no-args onSuccess() call.
+   */
+  public metricSuccess(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return JobBase.metricRule(this.onSuccess(), props);
+  }
+
+  /**
+   * Return a CloudWatch Metric indicating job failure.
+   *
+   * This metric is based on the Rule returned by no-args onFailure() call.
+   */
+  public metricFailure(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return JobBase.metricRule(this.onFailure(), props);
+  }
+
+  /**
+   * Return a CloudWatch Metric indicating job timeout.
+   *
+   * This metric is based on the Rule returned by no-args onTimeout() call.
+   */
+  public metricTimeout(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
+    return JobBase.metricRule(this.onTimeout(), props);
+  }
+
+  /**
+   * Creates a new rule for a transition into the input jobState or attempt to create-if-necessary and retrieve the default rule
+   * - A new rule is created but not cached if the id parameter is specified
+   * - A create/retrieve from cache scenario happens when no explicit id (and options) are not provided
+   * The reason is that the default rule is used by onSuccess, onFailure and onTimeout methods which are in turn used by metrics methods.
+   *
+   * @param jobState the job state
+   * @param id optional construct id
```

Comment:
I don't see any reason for this to be provided if we are doing caching already

Reply:
So
- the `onXxx` methods have to follow a specific signature for cloudwatch events
- that signature mandates the ability to set `id` and `options`
- The default `onSuccess()` et. al work just fine and are cached because of metrics methods reliance on them
- To be able to still abide to the expectations of `onXxx` methods, in that if u provide non-default parameters, caching is only restricted to the default parameterless versions
- This method `rule` is a place to centralize all that logic

Reply:
okay, understood why we're trying to cache. I think what we should do instead is move the caching to the metric methods that leverage events. they should have some ID that's particular to their state (EventMetricSucceeded, for example) and create/get an event with that ID (use [`ConstructNode.tryFindChild`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.ConstructNode.html#trywbrfindwbrchildid) instead of a map). the onXXX methods should all require an ID

Reply:
We could implement caching here but IMO it's out of scope for this class and would be inconsistent with the other constructs that provide `onEvent: Rule` methods. Happy to consider it as a feature for all such constructs in a separate PR!

Reply:
The main difference between this and others that have events, is that Glue does not have the basic CloudWatch Metrics that support basic monitoring/alarming. I think there's value here in providing those metrics and trying to make sure they're implemented efficiently and I'm not sure anything but caching would help with that.
There's an outstanding request to the Glue team to support some of those basic CloudWatch Metrics and if/when they support them, those methods can likely be changed to use the new metric-based implementations.
From experience, working with many glue jobs, we heavily rely on those metrics for alarming/dashboarding purposes and I'd hate for every invocation to end up creating an event rule.
What do you think?